content/Experimenting with Airflow to process S3 files - Mikaela Pisani.md

```diff
@@ -0,0 +1,343 @@
+
+To get the canonical user for s3: 
+```bash
+	aws s3api list-buckets --query Owner.Ioutput text
+```
+
+```python
+from airflow import DAG
+from airflow.operators.bash_operator import BashOperator
+from airflow.operators.s3_file_transform_operator import S3FileTransformOperator
+
+from datetime import datetime, timedelta
+
+
+default_args = {
+    "owner": "airflow",
+    "depends_on_past": False,
+    "start_date": datetime(2020, 9, 7),
+    "email": ["mikaela.pisani@rootstrap.com"],
+    "email_on_failure": False,
+    "email_on_retry": False,
+    "retries": 1,
+    "retry_delay": timedelta(minutes=5)
+}
+
+with DAG("s3_transformer", default_args=default_args, schedule_interval= '@once') as dag:
+
+    t1 = BashOperator(
+        task_id='bash_test',
+        bash_command='echo "hello, it should work" > s3_conn_test.txt'
+    )
+
+
+    transformer = S3FileTransformOperator(
+        task_id='ETL_medical_records',
+        description='cleans ETL_medical_records',
+        source_s3_key='s3://XXX/YYY/ZZZ.xml',
+        dest_s3_key='s3://XXX/YYY/WWW.xml',
+        replace=False,
+        transform_script='/usr/local/airflow/dags/scripts/transform.py',
+        source_aws_conn_id='s3_connection',
+        dest_aws_conn_id='s3_connection'
+    )
+
+    t1.set_upstream(transformer)
+```
+Change the parameters source\_s3\_key and dest\_s3\_key in the script, and copy the dag to dags folder.  
+
+```bash
+	 docker cp test_s3_file_transform_operator.py  docker-airflow_webserver_1:/usr/local/airflow/dags/
+```
+
+Create the script 'transform.py'
+
+```python
+	#!/usr/bin/python3
+	
+	
+	import sys
+	
+	input=sys.argv[1]
+	output=sys.argv[2]
+	
+	print("Starting data transformation...")
+	# DO SOMETHING
+	print("Completed data transformation!")
+```
+
+Copy it to the container: 
+
+```bash
+	 docker exec -ti docker-airflow_webserver_1 mkdir /usr/local/airflow/dags/scfipts/ & docker cp transform.py  docker-airflow_webserver_1:/usr/local/airflow/dags/scfipts/
+```
+
+In order to run this task, we will need to install some libraries in the containers, and then restart them: 
+
+```bash
+	docker exec -ti docker-airflow_worker_1 pip install boto3 boto botocore & docker exec -ti docker-airflow_scheduler_1 pip install boto3 boto botocore & docker exec -ti docker-airflow_webserver_1 pip install boto3 boto botocore
+```
+
+Restart the containers
+
+```bash
+	docker restart docker-airflow_worker_1 & docker restart docker-airflow_scheduler_1 & docker restart docker-airflow_webserver_1
+```
+
+![](images/s3_connection.png)
+
+It is important that the script that you set in the S3FileTransformOperator starts with **#!/usr/bin/python3 **in the case of python.      
+
+**Problem: if your script needs specific libraries to be installed (for example needs pandas), those are not installed in the worker, so when it executes the task gives you an error. For this problem there is not a clean solution, unless instead of celery you use KubernetesExecutor.**
```

Comment:
```suggestion
**Potential problem: if your script needs specific libraries to be installed (for example needs pandas), they are not installed in the worker. So, when it executes the task gives you an error. There is not a clean solution for this issue unless you use KubernetesExecutor instead of celery.**
```